main/src/library/MutHashMap.flix

```diff
@@ -0,0 +1,817 @@
+        }, m)
+    }
+
+    ///
+    /// Returns an iterator over key-value pairs in `m` in insertion order.
+    ///
+    pub def iterator(rc: Region[r2], m: MutHashMap[k, v, r]): Iterator[(k, v), r2 + r, r2] \ {r, r2} =
+        let curr = Ref.fresh(rc, m->head);
+        let next = () -> match Ref.get(curr) {
+            case None => None
+            case Some(entry) =>
+                Ref.put(Entry.getOrderNext(entry), curr);
+                Some((Entry.getKey(entry), Entry.getValue(entry)))
+        };
+        Iterator.iterate(rc, next)
+
+    ///
+    /// Returns an iterator over keys in `m` in insertion order.
+    ///
+    pub def iteratorKeys(rc: Region[r2], m: MutHashMap[k, v, r]): Iterator[k, r2 + r, r2] \ {r, r2} =
+        iterator(rc, m) |> Iterator.map(fst)
+
+    ///
+    /// Returns an iterator over values in `m` in insertion order.
+    ///
+    pub def iteratorValues(rc: Region[r2], m: MutHashMap[k, v, r]): Iterator[v, r2 + r, r2] \ {r, r2} =
+        iterator(rc, m) |> Iterator.map(snd)
+
+    ///
+    /// Returns a new mutable hash map with values transformed by function `f`.
+    ///
+    /// Keys and insertion order are preserved.
+    ///
+    pub def map(rc: Region[r2], f: v1 -> v2 \ ef, m: MutHashMap[k, v1, r1]): MutHashMap[k, v2, r2] \ {ef, r1, r2} with Eq[k], Hash[k] =
+        let result = emptyWithCapacity(rc, size(m));
+        forEach((k, v) -> put(k, f(v), result), m);
+        result
+
+    ///
+    /// Returns a new mutable hash map with values transformed by function `f` with access to keys.
+    ///
+    /// Keys and insertion order are preserved.
+    ///
+    pub def mapWithKey(rc: Region[r2], f: (k, v1) -> v2 \ ef, m: MutHashMap[k, v1, r1]): MutHashMap[k, v2, r2] \ {ef, r1, r2} with Eq[k], Hash[k] =
+        let result = emptyWithCapacity(rc, size(m));
+        forEach((k, v) -> put(k, f(k, v), result), m);
+        result
+
+    ///
+    /// Transforms all values in `m` in-place using function `f`.
+    ///
+    pub def transform(f: v -> v \ ef, m: MutHashMap[k, v, r]): Unit \ {ef, r} =
+        def loop(curr) = match curr {
+            case None => ()
+            case Some(entry) =>
+                Entry.setValue(f(Entry.getValue(entry)), entry);
+                loop(Entry.getOrderNext(entry))
+        };
+        loop(m->head)
+
+    ///
+    /// Transforms all values in `m` in-place using function `f` with access to keys.
+    ///
+    pub def transformWithKey(f: (k, v) -> v \ ef, m: MutHashMap[k, v, r]): Unit \ {ef, r} =
+        def loop(curr) = match curr {
+            case None => ()
+            case Some(entry) =>
+                Entry.setValue(f(Entry.getKey(entry), Entry.getValue(entry)), entry);
+                loop(Entry.getOrderNext(entry))
+        };
+        loop(m->head)
+
+    ///
+    /// Removes all entries from `m` where `f(value)` returns false.
+    ///
+    pub def refine(f: v -> Bool \ ef, m: MutHashMap[k, v, r]): Unit \ {ef, r} with Eq[k], Hash[k] =
```

Comment:
Should be called `filter`

Reply:
Actually not, because its return type is Unit.

Same for transform not being called map.

You can check MutMap API.